### PR TITLE
Use pattern matching to avoid strange type annotations

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Addresses https://github.com/flutter/flutter/pull/131640#discussion_r1284861384

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -1917,16 +1917,14 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
...
-      final TextPosition textPosition = _textPainter.getPositionForOffset(effectivePosition);
-      final Object? span = textSpan.getSpanForPosition(textPosition);
-      if (span is HitTestTarget) {
+    final Offset effectivePosition = position - _paintOffset;
+    switch (textSpan?.getSpanForPosition(_textPainter.getPositionForOffset(effectivePosition))) {
+      case final HitTestTarget span:
         result.add(HitTestEntry(span));
         return true;
-      }
+      case _:
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/editable.dart:1925]: is there a different between this and `default`?

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/editable.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:1925]: I think here they are the same.

Comment (Hixie) [reply packages/flutter/lib/src/rendering/editable.dart:1925]: in general i recommend leaning away from `_` where possible, fwiw. It's harder to understand at a glance. (Not sure if `default` works here though.)

**Comment by Hixie** [comment]
test-exempt: code refactor with no semantic change

File: packages/flutter/lib/src/rendering/editable.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:1925]: > Not sure if default works here though.

`default` does work here I believe.  

> It's harder to understand

How so? Wildcards can be used with other patterns (e.g. `case _ when x > 0 : `) and in other patterns (e.g. `[_, a, b, _]`), so they're more powerful than `default` and in switch expressions I think `default` is not allowed (e.g.  `switch(num) { 42 => true, default => false };` is not valid). So using `_` syntax in switch statements to catch all is more consistent with switch expressions I feel?

**Pull request merged**